File: code/packages/elixir/starlark_vm/lib/starlark_vm/vm.ex
defmodule CodingAdventures.StarlarkVm.Vm do
  @moduledoc """
  Starlark VM Factory — Creates configured GenericVM instances for Starlark.

  ## Chapter 1: The Full Pipeline

  This module ties everything together. The `create_starlark_vm/0` factory
  creates a `GenericVM` that's fully configured for Starlark execution:

  1. All 46+ opcodes have registered handlers.
  2. All 23 built-in functions are registered.
  3. Starlark-specific restrictions are configured (recursion limits, etc.).

  The `execute_starlark/1` convenience function goes even further: it takes
  Starlark source code as a string, compiles it, and executes it in one call.

  ## Chapter 2: How to Use

  **Quick start — one call does everything:**

      result = StarlarkVm.execute_starlark("x = 1 + 2\\nprint(x)\\n")
      result.variables["x"]  #=> 3
      result.output           #=> ["3"]

  **Step by step — for more control:**

      code = StarlarkAstToBytecodeCompiler.compile_starlark("x = 1 + 2\\n")
      vm = StarlarkVm.create_starlark_vm()
      {traces, vm} = GenericVM.execute(vm, code)
      vm.variables["x"]  #=> 3
  """

  alias CodingAdventures.VirtualMachine.GenericVM
  alias CodingAdventures.StarlarkAstToBytecodeCompiler.Opcodes, as: Op
  alias CodingAdventures.StarlarkVm.Handlers
  alias CodingAdventures.StarlarkVm.Builtins

  # ===========================================================================
  # VM Factory
  # ===========================================================================

  @doc """
  Create a `GenericVM` fully configured for Starlark execution.

  This is the main factory function. It:
  1. Creates a fresh GenericVM.
  2. Registers all 46+ Starlark opcode handlers.
  3. Registers all 23 Starlark built-in functions.
  4. Configures Starlark-specific restrictions.

  ## Options

  - `:max_recursion_depth` — Maximum call stack depth. Default 200.
  - `:frozen` — Whether to start in frozen mode. Default false.

  ## Example

      vm = Vm.create_starlark_vm()
      code = compile_starlark("x = 42\\n")
      {traces, vm} = GenericVM.execute(vm, code)
      vm.variables["x"]  #=> 42
  """
  def create_starlark_vm(opts \\ []) do
    max_depth = Keyword.get(opts, :max_recursion_depth, 200)
    frozen = Keyword.get(opts, :frozen, false)

    vm = GenericVM.new()

    # -- Register all opcode handlers --

    # Stack operations
    vm = GenericVM.register_opcode(vm, Op.load_const(), &Handlers.handle_load_const/3)
    vm = GenericVM.register_opcode(vm, Op.pop(), &Handlers.handle_pop/3)
    vm = GenericVM.register_opcode(vm, Op.dup(), &Handlers.handle_dup/3)
    vm = GenericVM.register_opcode(vm, Op.load_none(), &Handlers.handle_load_none/3)
    vm = GenericVM.register_opcode(vm, Op.load_true(), &Handlers.handle_load_true/3)
    vm = GenericVM.register_opcode(vm, Op.load_false(), &Handlers.handle_load_false/3)

    # Variable operations
    vm = GenericVM.register_opcode(vm, Op.store_name(), &Handlers.handle_store_name/3)
    vm = GenericVM.register_opcode(vm, Op.load_name(), &Handlers.handle_load_name/3)
    vm = GenericVM.register_opcode(vm, Op.store_local(), &Handlers.handle_store_local/3)
    vm = GenericVM.register_opcode(vm, Op.load_local(), &Handlers.handle_load_local/3)
    vm = GenericVM.register_opcode(vm, Op.store_closure(), &Handlers.handle_store_closure/3)
    vm = GenericVM.register_opcode(vm, Op.load_closure(), &Handlers.handle_load_closure/3)

    # Arithmetic
    vm = GenericVM.register_opcode(vm, Op.add(), &Handlers.handle_add/3)
    vm = GenericVM.register_opcode(vm, Op.sub(), &Handlers.handle_sub/3)
    vm = GenericVM.register_opcode(vm, Op.mul(), &Handlers.handle_mul/3)
    vm = GenericVM.register_opcode(vm, Op.div_op(), &Handlers.handle_div/3)
    vm = GenericVM.register_opcode(vm, Op.floor_div(), &Handlers.handle_floor_div/3)
    vm = GenericVM.register_opcode(vm, Op.mod(), &Handlers.handle_mod/3)
    vm = GenericVM.register_opcode(vm, Op.power(), &Handlers.handle_power/3)
    vm = GenericVM.register_opcode(vm, Op.negate(), &Handlers.handle_negate/3)
    vm = GenericVM.register_opcode(vm, Op.bit_and(), &Handlers.handle_bit_and/3)
    vm = GenericVM.register_opcode(vm, Op.bit_or(), &Handlers.handle_bit_or/3)
    vm = GenericVM.register_opcode(vm, Op.bit_xor(), &Handlers.handle_bit_xor/3)
    vm = GenericVM.register_opcode(vm, Op.bit_not(), &Handlers.handle_bit_not/3)
    vm = GenericVM.register_opcode(vm, Op.lshift(), &Handlers.handle_lshift/3)
    vm = GenericVM.register_opcode(vm, Op.rshift(), &Handlers.handle_rshift/3)

    # Comparisons
    vm = GenericVM.register_opcode(vm, Op.cmp_eq(), &Handlers.handle_cmp_eq/3)
    vm = GenericVM.register_opcode(vm, Op.cmp_ne(), &Handlers.handle_cmp_ne/3)
    vm = GenericVM.register_opcode(vm, Op.cmp_lt(), &Handlers.handle_cmp_lt/3)
    vm = GenericVM.register_opcode(vm, Op.cmp_gt(), &Handlers.handle_cmp_gt/3)
    vm = GenericVM.register_opcode(vm, Op.cmp_le(), &Handlers.handle_cmp_le/3)
    vm = GenericVM.register_opcode(vm, Op.cmp_ge(), &Handlers.handle_cmp_ge/3)
    vm = GenericVM.register_opcode(vm, Op.cmp_in(), &Handlers.handle_cmp_in/3)
    vm = GenericVM.register_opcode(vm, Op.cmp_not_in(), &Handlers.handle_cmp_not_in/3)

    # Boolean
    vm = GenericVM.register_opcode(vm, Op.logical_not(), &Handlers.handle_not/3)

    # Control flow
    vm = GenericVM.register_opcode(vm, Op.jump(), &Handlers.handle_jump/3)
    vm = GenericVM.register_opcode(vm, Op.jump_if_false(), &Handlers.handle_jump_if_false/3)
    vm = GenericVM.register_opcode(vm, Op.jump_if_true(), &Handlers.handle_jump_if_true/3)
    vm = GenericVM.register_opcode(vm, Op.jump_if_false_or_pop(), &Handlers.handle_jump_if_false_or_pop/3)
    vm = GenericVM.register_opcode(vm, Op.jump_if_true_or_pop(), &Handlers.handle_jump_if_true_or_pop/3)

    # Functions
    vm = GenericVM.register_opcode(vm, Op.make_function(), &Handlers.handle_make_function/3)
    vm = GenericVM.register_opcode(vm, Op.call_function(), &Handlers.handle_call_function/3)
    vm = GenericVM.register_opcode(vm, Op.call_function_kw(), &Handlers.handle_call_function_kw/3)
    vm = GenericVM.register_opcode(vm, Op.return_op(), &Handlers.handle_return/3)

    # Collections
    vm = GenericVM.register_opcode(vm, Op.build_list(), &Handlers.handle_build_list/3)
    vm = GenericVM.register_opcode(vm, Op.build_dict(), &Handlers.handle_build_dict/3)
    vm = GenericVM.register_opcode(vm, Op.build_tuple(), &Handlers.handle_build_tuple/3)
    vm = GenericVM.register_opcode(vm, Op.list_append(), &Handlers.handle_list_append/3)
    vm = GenericVM.register_opcode(vm, Op.dict_set(), &Handlers.handle_dict_set/3)

    # Subscript & attribute
    vm = GenericVM.register_opcode(vm, Op.load_subscript(), &Handlers.handle_load_subscript/3)
    vm = GenericVM.register_opcode(vm, Op.store_subscript(), &Handlers.handle_store_subscript/3)
    vm = GenericVM.register_opcode(vm, Op.load_attr(), &Handlers.handle_load_attr/3)
    vm = GenericVM.register_opcode(vm, Op.store_attr(), &Handlers.handle_store_attr/3)
    vm = GenericVM.register_opcode(vm, Op.load_slice(), &Handlers.handle_load_slice/3)

    # Iteration
    vm = GenericVM.register_opcode(vm, Op.get_iter(), &Handlers.handle_get_iter/3)
    vm = GenericVM.register_opcode(vm, Op.for_iter(), &Handlers.handle_for_iter/3)
    vm = GenericVM.register_opcode(vm, Op.unpack_sequence(), &Handlers.handle_unpack_sequence/3)

    # Module
    vm = GenericVM.register_opcode(vm, Op.load_module(), &Handlers.handle_load_module/3)
    vm = GenericVM.register_opcode(vm, Op.import_from(), &Handlers.handle_import_from/3)

    # I/O
    vm = GenericVM.register_opcode(vm, Op.print_op(), &Handlers.handle_print/3)

    # VM control
    vm = GenericVM.register_opcode(vm, Op.halt(), &Handlers.handle_halt/3)

    # -- Register built-in functions --
    vm = Enum.reduce(Builtins.get_all_builtins(), vm, fn {name, impl}, acc_vm ->
      GenericVM.register_builtin(acc_vm, name, impl)
    end)

    # -- Configure restrictions --
    vm = GenericVM.set_max_recursion_depth(vm, max_depth)
    vm = if frozen, do: GenericVM.set_frozen(vm, true), else: vm

    vm
  end

  # ===========================================================================
  # Convenience Functions
  # ===========================================================================

  @doc """
  Compile and execute Starlark source code in one call.

  This is the highest-level API. Pass in Starlark source code,
  get back the execution result with variables, output, and traces.

  ## Parameters

  - `source` — Starlark source code. Should end with a newline.

  ## Returns

  A `%StarlarkResult{}` struct with variables, output, and traces.

  ## Example

      result = Vm.execute_starlark("x = 1 + 2\\nprint(x)\\n")
      result.variables["x"]  #=> 3
      result.output           #=> ["3"]
  """
  def execute_starlark(source) when is_binary(source) do
    alias CodingAdventures.StarlarkAstToBytecodeCompiler.Compiler

    code = Compiler.compile_starlark(source)
    vm = create_starlark_vm()
    {traces, vm} = GenericVM.execute(vm, code)

    %Handlers.StarlarkResult{
      variables: vm.variables,
      output: Enum.reverse(vm.output),
      traces: traces
    }
  end
end

File: code/packages/python/starlark-vm/src/starlark_vm/vm.py
"""Starlark VM — A complete Starlark bytecode interpreter.

==========================================================================
Chapter 1: The Full Pipeline
==========================================================================

This module ties everything together. The ``create_starlark_vm()`` factory
creates a ``GenericVM`` that's fully configured for Starlark execution:

1. All ~50 opcodes have registered handlers.
2. All ~25 built-in functions are registered.
3. Starlark-specific restrictions are configured (recursion limits, etc.).

The ``execute_starlark()`` convenience function goes even further: it takes
Starlark source code as a string, compiles it, and executes it in one call.

==========================================================================
Chapter 2: How to Use
==========================================================================

**Quick start — one call does everything:**

    result = execute_starlark("x = 1 + 2\\nprint(x)\\n")
    print(result.variables["x"])  # 3
    print(result.output)          # ["3"]

**Step by step — for more control:**

    from starlark_ast_to_bytecode_compiler import compile_starlark
    from starlark_vm import create_starlark_vm

    # Compile
    code = compile_starlark("x = 1 + 2\\n")

    # Execute
    vm = create_starlark_vm()
    traces = vm.execute(code)

    # Inspect
    print(vm.variables["x"])  # 3
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from virtual_machine import GenericVM, VMTrace

from starlark_ast_to_bytecode_compiler.opcodes import Op
from starlark_vm.builtins import get_all_builtins
from starlark_vm.handlers import (
    handle_add,
    handle_bit_and,
    handle_bit_not,
    handle_bit_or,
    handle_bit_xor,
    handle_build_dict,
    handle_build_list,
    handle_build_tuple,
    handle_call_function,
    handle_call_function_kw,
    handle_cmp_eq,
    handle_cmp_ge,
    handle_cmp_gt,
    handle_cmp_in,
    handle_cmp_le,
    handle_cmp_lt,
    handle_cmp_ne,
    handle_cmp_not_in,
    handle_dict_set,
    handle_div,
    handle_dup,
    handle_floor_div,
    handle_for_iter,
    handle_get_iter,
    handle_halt,
    handle_import_from,
    handle_jump,
    handle_jump_if_false,
    handle_jump_if_false_or_pop,
    handle_jump_if_true,
    handle_jump_if_true_or_pop,
    handle_list_append,
    handle_load_attr,
    handle_load_closure,
    handle_load_const,
    handle_load_false,
    handle_load_local,
    handle_load_module,
    handle_load_name,
    handle_load_none,
    handle_load_slice,
    handle_load_subscript,
    handle_load_true,
    handle_lshift,
    handle_make_function,
    handle_mod,
    handle_mul,
    handle_negate,
    handle_not,
    handle_pop,
    handle_power,
    handle_print,
    handle_return,
    handle_rshift,
    handle_store_attr,
    handle_store_closure,
    handle_store_local,
    handle_store_name,
    handle_store_subscript,
    handle_sub,
    handle_unpack_sequence,
)


# =========================================================================
# VM Factory
# =========================================================================


def create_starlark_vm(
    max_recursion_depth: int = 200,
    frozen: bool = False,
) -> GenericVM:
    """Create a ``GenericVM`` fully configured for Starlark execution.

    This is the main factory function. It:
    1. Creates a fresh GenericVM.
    2. Registers all ~50 Starlark opcode handlers.
    3. Registers all ~25 Starlark built-in functions.
    4. Configures Starlark-specific restrictions.

    Parameters
    ----------
    max_recursion_depth : int
        Maximum call stack depth. Default 200. Set to 0 to completely
        forbid function calls (strict Starlark mode).
    frozen : bool
        Whether to start in frozen mode (no mutations allowed).

    Returns
    -------
    GenericVM
        A VM ready to execute Starlark bytecode.

    Example::

        vm = create_starlark_vm()
        code = compile_starlark("x = 42\\n")
        vm.execute(code)
        assert vm.variables["x"] == 42
    """
    vm = GenericVM()

    # -- Register all opcode handlers --

    # Stack operations
    vm.register_opcode(Op.LOAD_CONST, handle_load_const)
    vm.register_opcode(Op.POP, handle_pop)
    vm.register_opcode(Op.DUP, handle_dup)
    vm.register_opcode(Op.LOAD_NONE, handle_load_none)
    vm.register_opcode(Op.LOAD_TRUE, handle_load_true)
    vm.register_opcode(Op.LOAD_FALSE, handle_load_false)

    # Variable operations
    vm.register_opcode(Op.STORE_NAME, handle_store_name)
    vm.register_opcode(Op.LOAD_NAME, handle_load_name)
    vm.register_opcode(Op.STORE_LOCAL, handle_store_local)
    vm.register_opcode(Op.LOAD_LOCAL, handle_load_local)
    vm.register_opcode(Op.STORE_CLOSURE, handle_store_closure)
    vm.register_opcode(Op.LOAD_CLOSURE, handle_load_closure)

    # Arithmetic
    vm.register_opcode(Op.ADD, handle_add)
    vm.register_opcode(Op.SUB, handle_sub)
    vm.register_opcode(Op.MUL, handle_mul)
    vm.register_opcode(Op.DIV, handle_div)
    vm.register_opcode(Op.FLOOR_DIV, handle_floor_div)
    vm.register_opcode(Op.MOD, handle_mod)
    vm.register_opcode(Op.POWER, handle_power)
    vm.register_opcode(Op.NEGATE, handle_negate)
    vm.register_opcode(Op.BIT_AND, handle_bit_and)
    vm.register_opcode(Op.BIT_OR, handle_bit_or)
    vm.register_opcode(Op.BIT_XOR, handle_bit_xor)
    vm.register_opcode(Op.BIT_NOT, handle_bit_not)
    vm.register_opcode(Op.LSHIFT, handle_lshift)
    vm.register_opcode(Op.RSHIFT, handle_rshift)

    # Comparisons
    vm.register_opcode(Op.CMP_EQ, handle_cmp_eq)
    vm.register_opcode(Op.CMP_NE, handle_cmp_ne)
    vm.register_opcode(Op.CMP_LT, handle_cmp_lt)
    vm.register_opcode(Op.CMP_GT, handle_cmp_gt)
    vm.register_opcode(Op.CMP_LE, handle_cmp_le)
    vm.register_opcode(Op.CMP_GE, handle_cmp_ge)
    vm.register_opcode(Op.CMP_IN, handle_cmp_in)
    vm.register_opcode(Op.CMP_NOT_IN, handle_cmp_not_in)

    # Boolean
    vm.register_opcode(Op.NOT, handle_not)

    # Control flow
    vm.register_opcode(Op.JUMP, handle_jump)
    vm.register_opcode(Op.JUMP_IF_FALSE, handle_jump_if_false)
    vm.register_opcode(Op.JUMP_IF_TRUE, handle_jump_if_true)
    vm.register_opcode(Op.JUMP_IF_FALSE_OR_POP, handle_jump_if_false_or_pop)
    vm.register_opcode(Op.JUMP_IF_TRUE_OR_POP, handle_jump_if_true_or_pop)

    # Functions
    vm.register_opcode(Op.MAKE_FUNCTION, handle_make_function)
    vm.register_opcode(Op.CALL_FUNCTION, handle_call_function)
    vm.register_opcode(Op.CALL_FUNCTION_KW, handle_call_function_kw)
    vm.register_opcode(Op.RETURN, handle_return)

    # Collections
    vm.register_opcode(Op.BUILD_LIST, handle_build_list)
    vm.register_opcode(Op.BUILD_DICT, handle_build_dict)
    vm.register_opcode(Op.BUILD_TUPLE, handle_build_tuple)
    vm.register_opcode(Op.LIST_APPEND, handle_list_append)
    vm.register_opcode(Op.DICT_SET, handle_dict_set)

    # Subscript & attribute
    vm.register_opcode(Op.LOAD_SUBSCRIPT, handle_load_subscript)
    vm.register_opcode(Op.STORE_SUBSCRIPT, handle_store_subscript)
    vm.register_opcode(Op.LOAD_ATTR, handle_load_attr)
    vm.register_opcode(Op.STORE_ATTR, handle_store_attr)
    vm.register_opcode(Op.LOAD_SLICE, handle_load_slice)

    # Iteration
    vm.register_opcode(Op.GET_ITER, handle_get_iter)
    vm.register_opcode(Op.FOR_ITER, handle_for_iter)
    vm.register_opcode(Op.UNPACK_SEQUENCE, handle_unpack_sequence)

    # Module
    vm.register_opcode(Op.LOAD_MODULE, handle_load_module)
    vm.register_opcode(Op.IMPORT_FROM, handle_import_from)

    # I/O
    vm.register_opcode(Op.PRINT, handle_print)

    # VM control
    vm.register_opcode(Op.HALT, handle_halt)

    # -- Register built-in functions --
    for name, impl in get_all_builtins().items():
        vm.register_builtin(name, impl)

    # Override print() with a closure that captures output to the VM.
    # The default builtin_print returns None without side effects because
    # builtins don't have access to the VM instance. This closure fixes
    # that by writing to vm.output directly, matching the PRINT opcode's
    # behavior for calls made via ``print("hello")``.
    def _print_with_capture(args: list) -> None:
        output_str = " ".join(str(a) for a in args)
        vm.output.append(output_str)
        return None

    vm.register_builtin("print", _print_with_capture)

    # -- Configure restrictions --
    vm.set_max_recursion_depth(max_recursion_depth)
    if frozen:
        vm.set_frozen(True)

    return vm


# =========================================================================
# Execution Result
# =========================================================================


@dataclass
class StarlarkResult:
    """The result of executing a Starlark program.

    Contains all the information about the execution:
    - variables: the final state of all named variables
    - output: captured print output
    - traces: step-by-step execution trace (for debugging)
    """

    variables: dict[str, Any]
    """Final variable state after execution."""

    output: list[str]
    """Captured print output, one entry per print() call."""

    traces: list[VMTrace]
    """Step-by-step execution trace."""


# =========================================================================
# Convenience Functions
# =========================================================================


def execute_starlark(source: str) -> StarlarkResult:
    """Compile and execute Starlark source code in one call.

    This is the highest-level API. Pass in Starlark source code,
    get back the execution result with variables, output, and traces.

    Parameters
    ----------
    source : str
        Starlark source code. Should end with a newline.

    Returns
    -------
    StarlarkResult
        The execution result.

    Example::

        result = execute_starlark("x = 1 + 2\\nprint(x)\\n")
        assert result.variables["x"] == 3
        assert result.output == ["3"]
    """
    from starlark_ast_to_bytecode_compiler import compile_starlark

    code = compile_starlark(source)
    vm = create_starlark_vm()
    traces = vm.execute(code)

    return StarlarkResult(
        variables=dict(vm.variables),
        output=list(vm.output),
        traces=traces,
    )
